accounts-db/src/file_io.rs

```diff
@@ -0,0 +1,151 @@
+//! File i/o helper functions.
+#[cfg(unix)]
+use std::os::unix::prelude::FileExt;
+use std::{fs::File, ops::Range};
+
+/// `buffer` contains `valid_bytes` of data at its end.
+/// Move those valid bytes to the beginning of `buffer`, then read from `offset` to fill the rest of `buffer`.
+/// Update `offset` for the next read and update `valid_bytes` to specify valid portion of `buffer`.
+/// `valid_len` is # of valid bytes in the file. This may be <= file length.
+pub fn read_more_buffer(
+    file: &File,
+    offset: &mut usize,
+    buffer: &mut [u8],
+    valid_bytes: &mut Range<usize>,
+    valid_len: usize,
+) -> std::io::Result<()> {
+    // copy remainder of `valid_bytes` into beginning of `buffer`.
+    // These bytes were left over from the last read but weren't enough for the next desired contiguous read chunk.
+    buffer.copy_within(valid_bytes.clone(), 0);
+
+    // read the rest of `buffer`
+    let bytes_read = read_buffer(file, *offset, &mut buffer[valid_bytes.len()..], valid_len)?;
+    *offset += bytes_read;
+    *valid_bytes = 0..(valid_bytes.len() + bytes_read);
+
+    Ok(())
+}
+
+#[cfg(unix)]
+/// Read, starting at `start_offset`, until `buffer` is full or we read past `valid_len`/eof.
+/// `valid_len` is # of valid bytes in the file. This may be <= file length.
+/// return # bytes read
+pub fn read_buffer(
+    file: &File,
+    start_offset: usize,
+    buffer: &mut [u8],
+    valid_len: usize,
+) -> std::io::Result<usize> {
+    let mut offset = start_offset;
+    let mut start_read = 0;
+    let mut total_bytes_read = 0;
+    if start_offset >= valid_len {
+        return Ok(0);
```

Comment:
caller doesn't have to track valid len of file/know how many bytes are valid in the file. This is how we know to stop reading. There are more bytes in the file, but they aren't valid.

Reply:
Yes I get the invalid padding stuff in files.

This function at the moment has two callers: get_stored_account_meta_callback and read_more_buffer.

In get_stored_account_meta_callback it's not clear to me that offset should ever be > self.len()? In which case I think it would be better to assert or handle the condition explicitly. Right now I think read_into_buffer will return Ok(0), then Self::get_type() will return None reading from the empty slice. I'd rather make the condition explicit.

read_more_buffer is called by BufferedReader::read. read is already asserting something, and it looks like like it could assert the offset > len condition too, or handle it.

I'm your guest in this area of the code, so ultimately up to you. But I personally prefer seeing the happy path clearly separated from the edge conditions, with the edge conditions isolated at the boundaries, so that the inner code can make stronger invariants and have less complexity.

Reply:
I can appreciate all of that. I believe what we have is correct. I think what you're discussing is a refactoring we could do afterwards. There are only 2 callers as you said. This feature is 3 months out from use and is cli opt-in atm.